public/app/features/alerting/unified/utils/time.ts

```diff
@@ -95,7 +95,68 @@ export function parsePrometheusDuration(duration: string): number {
   return totalDuration;
 }
 
-export const safeParseDurationstr = (duration: string): number => {
+export function formatPrometheusDuration(milliseconds: number): string {
+  const seconds = Math.floor(milliseconds / 1000);
+  const minutes = Math.floor(seconds / 60);
+  const hours = Math.floor(minutes / 60);
+  const days = Math.floor(hours / 24);
+  const weeks = Math.floor(days / 7);
+  const years = Math.floor(days / 365);
+
+  // we'll make an exception here for 0, 0ms seems a bit weird
+  if (milliseconds === 0) {
+    return '0s';
+  }
+
+  if (milliseconds < 1000) {
```

Comment:
Was thinking if there's a cleaner way to do it, something like the below - but, in trying to do this I probably ended up making it too smart for its own good, and harder to understand than the else/ifs

```ts
export function formatPrometheusDuration(milliseconds: number): string {
  const seconds = Math.floor(milliseconds / 1000);
  const minutes = Math.floor(seconds / 60);
  const hours = Math.floor(minutes / 60);
  const days = Math.floor(hours / 24);
  const weeks = Math.floor(days / 7);
  const years = Math.floor(days / 365);

  // we'll make an exception here for 0, 0ms seems a bit weird
  if (milliseconds === 0) {
    return '0s';
  }

  const allUnits: Array<[number, string]> = [
    [years, 'y'],
    [weeks % 52, 'w'],
    [(days % 365) - 7 * (weeks % 52), 'd'],
    [hours % 24, 'h'],
    [minutes % 60, 'm'],
    [seconds % 60, 's'],
    [milliseconds % 1000, 'ms'],
  ];

  // Find the largest non-zero unit that we'll have to show
  const largestUnitIndex = allUnits.findIndex(([time]) => {
    return time > 0;
  });

  // Are there any units below the largest leading unit that are non-zero? 
  // e.g. is the converted value of milliseconds an exact rounded value?
  const isRoundUnitValue = [...allUnits.slice(largestUnitIndex + 1)].some(([time]) => time > 0);

  const filtered = allUnits.filter(([time], index) => {
    if (index < largestUnitIndex) {
      return false;
    }
    return time > 0 || isRoundUnitValue;
  });

  return combineUnits(...filtered);
}
```

And turning `combineUnits` into the below, as it doesn't need to filter the `0` values:

```ts
function combineUnits(...units: Array<[number, string]>): string {
  return units.map((u) => u.join('')).join('');
}
```

Reply:
Clever solution! Let's go with that :)